Zend/zend_enum.c

```diff
@@ -0,0 +1,338 @@
+   | http://www.zend.com/license/2_00.txt.                                |
+   | If you did not receive a copy of the Zend license and are unable to  |
+   | obtain it through the world-wide-web, please send a note to          |
+   | license@zend.com so we can mail you a copy immediately.              |
+   +----------------------------------------------------------------------+
+   | Authors: Ilija Tovilo <ilutov@php.net>                               |
+   +----------------------------------------------------------------------+
+*/
+
+#include "zend.h"
+#include "zend_API.h"
+#include "zend_compile.h"
+#include "zend_interfaces_arginfo.h"
+
+#define ZEND_ENUM_PROPERTY_ERROR() \
+	zend_throw_error(NULL, "Enum properties are immutable")
+
+#define ZEND_ENUM_DISALLOW_MAGIC_METHOD(propertyName, methodName) \
+	do { \
+		if (ce->propertyName) { \
+			zend_error_noreturn(E_COMPILE_ERROR, "Enum may not include %s", methodName); \
+		} \
+	} while (0);
+
+static zend_object_handlers enum_handlers;
+
+zend_object *zend_enum_new(zval *result, zend_class_entry *ce, zval *case_name_zv, zval *scalar_zv)
+{
+	// Temporarily remove the ZEND_ACC_ENUM flag to allow instantiation
+	ce->ce_flags &= ~ZEND_ACC_ENUM;
+	object_init_ex(result, ce);
+	ce->ce_flags |= ZEND_ACC_ENUM;
+
+	zend_object *zobj = Z_OBJ_P(result);
+	zobj->handlers->write_property(zobj, ZSTR_KNOWN(ZEND_STR_NAME), case_name_zv, NULL);
+	if (scalar_zv != NULL) {
+		zobj->handlers->write_property(zobj, ZSTR_KNOWN(ZEND_STR_VALUE), scalar_zv, NULL);
+	}
+
+	zobj->handlers = &enum_handlers;
+
+	return zobj;
+}
+
+static void zend_verify_enum_properties(zend_class_entry *ce)
+{
+	zend_property_info *property_info;
+
+	ZEND_HASH_FOREACH_PTR(&ce->properties_info, property_info) {
+		if (zend_string_equals_literal(property_info->name, "name")) {
+			continue;
+		}
+		if (
+			ce->enum_scalar_type != IS_UNDEF
+			&& zend_string_equals_literal(property_info->name, "value")
+		) {
+			continue;
+		}
+		// FIXME: File/line number for traits?
+		zend_error_noreturn(E_COMPILE_ERROR, "Enum \"%s\" may not include member variables",
+			ZSTR_VAL(ce->name));
+	} ZEND_HASH_FOREACH_END();
+}
+
+static void zend_verify_enum_magic_methods(zend_class_entry *ce)
+{
+	// Only __get, __call and __invoke are allowed
+
+	ZEND_ENUM_DISALLOW_MAGIC_METHOD(constructor, "__construct");
+	ZEND_ENUM_DISALLOW_MAGIC_METHOD(destructor, "__destruct");
+	ZEND_ENUM_DISALLOW_MAGIC_METHOD(clone, "__clone");
+	ZEND_ENUM_DISALLOW_MAGIC_METHOD(__get, "__get");
+	ZEND_ENUM_DISALLOW_MAGIC_METHOD(__set, "__set");
+	ZEND_ENUM_DISALLOW_MAGIC_METHOD(__unset, "__unset");
+	ZEND_ENUM_DISALLOW_MAGIC_METHOD(__isset, "__isset");
+	ZEND_ENUM_DISALLOW_MAGIC_METHOD(__tostring, "__toString");
+	ZEND_ENUM_DISALLOW_MAGIC_METHOD(__debugInfo, "__debugInfo");
+	ZEND_ENUM_DISALLOW_MAGIC_METHOD(__serialize, "__serialize");
+	ZEND_ENUM_DISALLOW_MAGIC_METHOD(__unserialize, "__unserialize");
```

Comment:
We should possibly also forbid the implementation of `Serializable` somewhere?

Reply:
Right, due to the custom serialization for enums this has no effect anyway.